### ISSUE TITLE
[2.1] 2D rendering issue regarding windows focus

### ISSUE THREAD

**Issue by xsellier** [godotengine/godot#29518]
**Godot version:**
2.1.5

**OS/device including version:**
- Windows (only windows user reported it)
- NVidia GTX 1080Ti

**Issue description:**
- Most of the rendering is **blurry** when **godot** has the **focus**.
- But if you **unfocus** **godot**, everything gets **sharp** and **well rendered**.

**Minimal project + step to reproduce**
https://github.com/xsellier/godot-2.1-render-bug

**List of people who are having the same issue**
https://steamcommunity.com/app/726840/discussions/0/1651043958646410385/
https://steamcommunity.com/app/726840/discussions/0/1842440600605384185/ (look at the last message for more informations)
https://steamcommunity.com/app/726840/discussions/0/1797403972744401372/
https://steamcommunity.com/profiles/76561198141817886/recommended/726840/

**Comment by Calinou** [comment]
Godot 2.1 doesn't feature DynamicFont oversampling, so if you use the `2d` stretch mode, you will have to find another way to get crisp visuals. One way is to use an high "design resolution" in the Project Settings (by setting the width and height to `3840` and `2160`, for instance), then set the **Test Width** and **Test Height** to lower values when testing the project.

**Comment by xsellier** [comment]
Not related to font oversampling, but thank you.

**Comment by Calinou** [comment]
@xsellier The comment I wrote also applies to other 2D elements, as you basically need to ship higher-resolution assets than what you'd use at most resolutions.

**Comment by xsellier** [comment]
Please, read the issue, please do it:
https://github.com/binogure-studio/city-game-studio-i18n/issues/105

**Comment by Calinou** [comment]
I don't know what could be causing the visual output to change depending on whether the game window is focused or not.

That said, I can see a separate issue on the second screenshot: when you enable filter and mipmaps, Godot will downscale 2D elements using linear filtering, which doesn't look great for heavily downscaled images. [Godot 3.1 has an `Image` method to resize images with trilinear interpolation](https://github.com/godotengine/godot/pull/19313) (which usually looks better), but it can only be called manually. In other words, it can't be used to resize *all* 2D elements with trilinear interpolation.

**Comment by xsellier** [comment]
You are right, it _might_ be related to downscaling images. However, here is a sprite (so, not a UI element), 
On left, application is on background, and on the right application is on foreground:
![image](https://user-images.githubusercontent.com/6616005/58991773-19047c00-87b7-11e9-9b18-751d3f9fb5c7.png)

**Comment by xsellier** [comment]
Just to add a bit more informations about it, the same scene from Godot Engine
![image](https://user-images.githubusercontent.com/6616005/58992165-0e96b200-87b8-11e9-9b7c-f9b426b05513.png)

**Comment by xsellier** [comment]
And also, this happens only on Windows, I cannot reproduce it on Linux/MacOS.

**Comment by xsellier** [comment]
Here is another _tiny_ issue, I'm not sure that it is a different issue.

On the left, you can see that there is an issue with the rendering. On the right the image is properly rendered. Difference between those 2 images is the camera position. I moved it by 2 pixels on the right.
![image](https://user-images.githubusercontent.com/6616005/58992878-f7f15a80-87b9-11e9-8725-c87e6385e101.png)

**Comment by Xrayez** [comment]
@xsellier regarding your *tiny* issue, I think you could try to enable `rendering/quality/2d/use_pixel_snap` in project settings, so that sub-pixel camera position won't affect rendering, though I'm not sure whether it's present in 2.1.

I think zooming could also somehow interfere with this (mipmaps generation issue)?

**Comment by xsellier** [comment]
> @xsellier regarding your _tiny_ issue, I think you could try to enable `rendering/quality/2d/use_pixel_snap` in project settings, so that sub-pixel camera position won't affect rendering, though I'm not sure whether it's present in 2.1. If not, you could try to manually clamp camera position yourself to "integer" coordinates.
> 
> I think zooming could also somehow interfere with this (mipmaps generation issue)?

Nice suggestions. I tried with `pixel_snap` enabled, it fixes the _tiny_ issue. Since I did not want to use `pixel_snap` because it breaks some animations (stuttering), I replaced the `set_pos` function of the camera by something like:
```
func set_camera_pos(position):
  position.x = int(position.x)
  position.y = int(position.y)

  set_pos(position)
```

**Comment by xsellier** [comment]
About the main issue, 2D rendering, I've narrowed it to the following conditions:
- Windows (from 8 to 10)
- Screen resolution above 2560 x 1440
- Set Godot Engine to fullscreen
- Graphic card: NVidia

So this issue is not related to this windows issue: https://support.microsoft.com/en-us/help/2802799/display-resolutions-higher-than-2048x1152-may-not-display-properly-on

So this is an actual screenshot of the minimal project:
![image](https://user-images.githubusercontent.com/6616005/59040445-ad68f000-8844-11e9-9705-453e0f1f46fe.png)

As you can see, the image resolution is 2560 x 1440 but it shows only 2048 x 1152.
The screenshot isn't blurry because the issue cannot be _screenshot_, meaning if you want to see the issue you need to take an actual picture of your screen.

**Comment by Calinou** [comment]
If you're using scaling above 100% in Windows, try making the Godot executable DPI-aware as explained in https://github.com/godotengine/godot/issues/24922#issuecomment-455479830

**Comment by xsellier** [comment]
@Calinou Thank you, since in v2.1.* High DPI is not implemented for Windows, I had to cherry-pick this commit: https://github.com/godotengine/godot/commit/eb0f9651b3bb12b79db4ee4f8862a3be16a11ea5

Backporting it to v2.1 branch fixes the issue

**Comment by akien-mga** [comment]
Fixed by #29550.
